File: bindings/vbnet/tests/EveraniumItb.VisualBasic.Tests/RekeyTests.vb
' Init -> Rekey -> Open receiver with the rotated blob -> round trip.

Imports System.Text
Imports Everanium.Itb.VisualBasic
Imports Xunit

Public Class RekeyTests

    <Fact>
    Public Sub RekeyRoundTrip()
        Using sender As Pipeline = Pipeline.Init("singlemsg-triple-mac-v1")
            Dim blobBefore As Byte() = sender.Blob

            Dim perm(31) As Byte
            Array.Fill(perm, CByte(&H11))
            Dim wrap(31) As Byte
            Array.Fill(wrap, CByte(&H22))
            sender.Rekey(perm, wrap)
            Assert.False(sender.Blob.SequenceEqual(blobBefore))

            Using receiver As Pipeline = Pipeline.Open("singlemsg-triple-mac-v1", sender.Blob)
                Dim plain As Byte() = Encoding.UTF8.GetBytes("post-rekey payload")
                Dim wire As Byte() = sender.EncryptMessage(plain)
                Assert.Equal(Of Byte)(plain, receiver.DecryptMessage(wire))
            End Using
        End Using
    End Sub
End Class

File: bindings/fsharp/tests/Everanium.LibItb3.FSharp.Tests/RekeyTests.fs
// init -> rekey -> load receiver from the rotated blob -> round
// trip.

module Everanium.Itb3.FSharp.Tests.RekeyTests

open System.Text
open Xunit
open Everanium.Itb3.FSharp
open Everanium.Itb3.FSharp.Tests.TestSupport

[<Fact>]
let ``rekey round trip`` () =
    use sender = unwrap (Pipeline.init "singlemsg-triple-mac-v1" Opts.empty)
    let blobBefore = unwrap (Pipeline.save sender)

    let perm = Array.create 32 0x11uy
    let wrap = Array.create 32 0x22uy
    unwrap (Pipeline.rekey sender perm wrap) |> ignore
    Assert.NotEqual<byte[]>(blobBefore, unwrap (Pipeline.save sender))

    use receiver = unwrap (Pipeline.load (unwrap (Pipeline.save sender)))
    let plain = Encoding.UTF8.GetBytes "post-rekey payload"
    let wire = unwrap (Pipeline.encryptMessage sender plain)
    Assert.Equal<byte[]>(plain, unwrap (Pipeline.decryptMessage receiver wire))
